security-csrf/src/test/java/io/micronaut/security/csrf/repository/CsrfDoubleSubmitCookiePatternTest.java

```diff
@@ -0,0 +1,175 @@
+@Property(name = "micronaut.security.csrf.signature-key", value = "pleaseChangeThisSecretForANewOnekoqQ-EstJQLr_T-1qS0gZH75aKtMN3Yj0iPS4hcgUuTwjAzZr1Z9CAow")
+@Property(name = "micronaut.security.redirect.enabled", value = StringUtils.FALSE)
+@Property(name = "micronaut.security.csrf.filter.regex-pattern", value = "^(?!\\/login).*$")
+@Property(name = "spec.name", value = "CsrfDoubleSubmitCookiePatternTest")
+@MicronautTest
+class CsrfDoubleSubmitCookiePatternTest {
+    public static final String FIX_SESSION_ID = "123456789";
+
+    @Test
+    void loginSavesACsrfTokenInCookie(@Client("/") HttpClient httpClient,
+                                      CsrfConfiguration csrfConfiguration) throws NoSuchAlgorithmException, InvalidKeyException {
+        BlockingHttpClient client = httpClient.toBlocking();
+
+        HttpRequest<?> loginRequest = HttpRequest.POST("/login",Map.of("username",  "sherlock", "password", "password"))
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED_TYPE);
+
+        HttpResponse<?> loginRsp = assertDoesNotThrow(() -> client.exchange(loginRequest));
+        assertEquals(HttpStatus.OK, loginRsp.getStatus());
+        Optional<Cookie> cookieJwtOptional = loginRsp.getCookie("JWT");
+        assertTrue(cookieJwtOptional.isPresent());
+        Cookie cookieJwt = cookieJwtOptional.get();
+        String csrfTokenCookieName = "__Host-csrfToken";
+        Optional<Cookie> cookieCsrfTokenOptional = loginRsp.getCookie(csrfTokenCookieName);
+        assertTrue(cookieCsrfTokenOptional.isPresent());
+        Cookie cookieCsrfToken = cookieCsrfTokenOptional.get();
+
+        // CSRF Only in the cookie, not in the request headers or field, request is denied
+        assertDenied(client, cookieJwt.getValue(), csrfTokenCookieName,  new PasswordChange("sherlock", "evil"), cookieCsrfToken.getValue());
+
+        // CSRF Token in request and in cookie don't match, request is unauthorized
+        String csrfToken = "abcdefg";
+        assertNotEquals(cookieCsrfToken.getValue(), csrfToken);
+        PasswordChangeForm formWithCsrfToken = new PasswordChangeForm("sherlock", "evil", csrfToken);
+        assertDenied(client, cookieJwt.getValue(), csrfTokenCookieName, formWithCsrfToken, cookieCsrfToken.getValue());
+
+        // CSRF Token with HMAC but not session id feed into HMAC calculation, request is unauthorized
+        String randomValue = "abcdefg";
+        String hmac = HMacUtils.base64EncodedHmacSha256(randomValue, csrfConfiguration.getSecretKey());
+        String csrfTokenCalculatedWithoutSessionId = hmac + "." + randomValue;
+        PasswordChangeForm body = new PasswordChangeForm("sherlock", "evil", csrfTokenCalculatedWithoutSessionId);
+        assertDenied(client, cookieJwt.getValue(), csrfTokenCookieName, body, csrfTokenCalculatedWithoutSessionId);
+
+        String message = FIX_SESSION_ID + "!" + randomValue;
```

Comment:
move into a static method in the generator that is VisibleForTesting

Reply:
I extracted the method https://github.com/micronaut-projects/micronaut-security/pull/1826/commits/36f45cdf124125ad907cd4872c531600ed96f4da I would preferred not to extract a one line method.

Reply:
ok, your call